// @progress/kendo-angular-grid/dist/es2015/filtering/menu/single-popup.service.js
import { PopupService } from '@progress/kendo-angular-popup';
import { Injectable, Renderer2, NgZone } from '@angular/core';
import { isPresent } from '../../utils';
import { PreventableEvent } from '../../common/preventable-event';
import { Subject } from 'rxjs/Subject';
import { LocalizationService } from '@progress/kendo-angular-l10n';
import { ScrollSyncService } from '../../scrolling/scroll-sync.service';
const contains = (node, predicate) => {
    while (node) {
        if (predicate(node)) {
            return true;
        }
        node = node.parentNode;
    }
    return false;
};
class CloseEvent extends PreventableEvent {
}
/**
 * @hidden
 */
export class SinglePopupService {
    constructor(popupService, renderer, ngZone, scrollSyncService, localization) {
        this.popupService = popupService;
        this.renderer = renderer;
        this.ngZone = ngZone;
        this.localization = localization;
        this.onClose = new Subject();
        this.scrollSubscription = scrollSyncService.changes.subscribe(() => this.destroy());
    }
    open(anchor, template, popupRef) {
        const toggle = isPresent(popupRef) && this.popupRef === popupRef;
        this.destroy();
        if (!toggle) {
            const direction = this.localization.rtl ? 'right' : 'left';
            this.popupRef = this.popupService.open({
                anchorAlign: { vertical: 'bottom', horizontal: direction },
                popupAlign: { vertical: 'top', horizontal: direction },
                anchor: anchor,
                popupClass: "k-grid-filter-popup",
                content: template,
                positionMode: "absolute"
            });
            this.popupViewPortLeaveSubscription = this.popupRef.popupAnchorViewportLeave
                .subscribe(() => this.destroy());
            this.attachClose(anchor);
        }
        return this.popupRef;
    }
    destroy() {
        if (this.popupRef) {
            this.detachClose();
            this.popupViewPortLeaveSubscription.unsubscribe();
            this.popupRef.close();
            this.popupRef = null;
        }
    }
    ngOnDestroy() {
        this.destroy();
        this.scrollSubscription.unsubscribe();
    }
    detachClose() {
        if (this.removeClick) {
            this.removeClick();
        }
    }
    attachClose(skipElement) {
        this.detachClose();
        this.ngZone.runOutsideAngular(() => this.removeClick = this.renderer.listen("document", "click", ({ target }) => {
            if (!contains(target, x => this.popupRef.popupElement === x || x === skipElement)) {
                const args = new CloseEvent();
                this.onClose.next(args);
                if (!args.isDefaultPrevented()) {
                    this.destroy();
                }
            }
        }));
    }
}
SinglePopupService.decorators = [
    { type: Injectable },
];
/** @nocollapse */
SinglePopupService.ctorParameters = () => [
    { type: PopupService, },
    { type: Renderer2, },
    { type: NgZone, },
    { type: ScrollSyncService, },
    { type: LocalizationService, },
];
