// @progress/kendo-angular-grid/dist/es/filtering/menu/single-popup.service.js
import * as tslib_1 from "tslib";
import { PopupService } from '@progress/kendo-angular-popup';
import { Injectable, Renderer2, NgZone } from '@angular/core';
import { isPresent } from '../../utils';
import { PreventableEvent } from '../../common/preventable-event';
import { Subject } from 'rxjs/Subject';
import { LocalizationService } from '@progress/kendo-angular-l10n';
import { ScrollSyncService } from '../../scrolling/scroll-sync.service';
var contains = function (node, predicate) {
    while (node) {
        if (predicate(node)) {
            return true;
        }
        node = node.parentNode;
    }
    return false;
};
var CloseEvent = /** @class */ (function (_super) {
    tslib_1.__extends(CloseEvent, _super);
    function CloseEvent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return CloseEvent;
}(PreventableEvent));
/**
 * @hidden
 */
var SinglePopupService = /** @class */ (function () {
    function SinglePopupService(popupService, renderer, ngZone, scrollSyncService, localization) {
        var _this = this;
        this.popupService = popupService;
        this.renderer = renderer;
        this.ngZone = ngZone;
        this.localization = localization;
        this.onClose = new Subject();
        this.scrollSubscription = scrollSyncService.changes.subscribe(function () { return _this.destroy(); });
    }
    SinglePopupService.prototype.open = function (anchor, template, popupRef) {
        var _this = this;
        var toggle = isPresent(popupRef) && this.popupRef === popupRef;
        this.destroy();
        if (!toggle) {
            var direction = this.localization.rtl ? 'right' : 'left';
            this.popupRef = this.popupService.open({
                anchorAlign: { vertical: 'bottom', horizontal: direction },
                popupAlign: { vertical: 'top', horizontal: direction },
                anchor: anchor,
                popupClass: "k-grid-filter-popup",
                content: template,
                positionMode: "absolute"
            });
            this.popupViewPortLeaveSubscription = this.popupRef.popupAnchorViewportLeave
                .subscribe(function () { return _this.destroy(); });
            this.attachClose(anchor);
        }
        return this.popupRef;
    };
    SinglePopupService.prototype.destroy = function () {
        if (this.popupRef) {
            this.detachClose();
            this.popupViewPortLeaveSubscription.unsubscribe();
            this.popupRef.close();
            this.popupRef = null;
        }
    };
    SinglePopupService.prototype.ngOnDestroy = function () {
        this.destroy();
        this.scrollSubscription.unsubscribe();
    };
    SinglePopupService.prototype.detachClose = function () {
        if (this.removeClick) {
            this.removeClick();
        }
    };
    SinglePopupService.prototype.attachClose = function (skipElement) {
        var _this = this;
        this.detachClose();
        this.ngZone.runOutsideAngular(function () {
            return _this.removeClick = _this.renderer.listen("document", "click", function (_a) {
                var target = _a.target;
                if (!contains(target, function (x) { return _this.popupRef.popupElement === x || x === skipElement; })) {
                    var args = new CloseEvent();
                    _this.onClose.next(args);
                    if (!args.isDefaultPrevented()) {
                        _this.destroy();
                    }
                }
            });
        });
    };
    SinglePopupService.decorators = [
        { type: Injectable },
    ];
    /** @nocollapse */
    SinglePopupService.ctorParameters = function () { return [
        { type: PopupService, },
        { type: Renderer2, },
        { type: NgZone, },
        { type: ScrollSyncService, },
        { type: LocalizationService, },
    ]; };
    return SinglePopupService;
}());
export { SinglePopupService };
